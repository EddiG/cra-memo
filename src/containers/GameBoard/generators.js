import shuffle from 'lodash/shuffle';
import flatten from 'lodash/flatten';
import times from 'lodash/times';
import random from 'lodash/random';
import first from 'lodash/first';
import {
  selectCardsQuantity,
  selectFrontImages,
  selectBackImages,
} from './selectors';

const getRandomItem = items => {
  const randomIndex = random(items.length - 1);
  return items[randomIndex];
};

export function generateCards(state, needShuffle = true) {
  const cardsQuantity = selectCardsQuantity(state);
  const frontImagesIds = selectFrontImages(state);
  const backImagesIds = selectBackImages(state);
  const cardsPairsQuantity = cardsQuantity / 2;
  const cardsPairs = times(cardsPairsQuantity, idx => {
    let frontImageId = '';
    if (frontImagesIds.length >= idx + 1) {
      frontImageId = frontImagesIds[idx];
    } else {
      frontImageId = getRandomItem(frontImagesIds);
    }
    const backImageId = first(backImagesIds);
    const cardFirstId = `card${idx + 1}`;
    const cardSecondId = `card${idx + 1 + cardsPairsQuantity}`;
    return [
      {
        id: cardFirstId,
        pairCardId: cardSecondId,
        isOpened: false,
        isGuessed: false,
        frontImageId,
        backImageId,
      },
      {
        id: cardSecondId,
        pairCardId: cardFirstId,
        isOpened: false,
        isGuessed: false,
        frontImageId,
        backImageId,
      },
    ];
  });
  const cards = flatten(cardsPairs);
  return needShuffle ? shuffle(cards) : cards;
}

export default generateCards;
