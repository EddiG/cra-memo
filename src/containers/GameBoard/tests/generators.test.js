import { generateCards } from '../generators';

const state = {
  frontImages: {
    byId: {
      frontImage1: {
        id: 'frontImage1',
        src: 'source1',
      },
      frontImage2: {
        id: 'frontImage2',
        src: 'source2',
      },
    },
    allIds: ['frontImage1', 'frontImage2'],
  },
  backImages: {
    byId: {
      backImage1: {
        id: 'backImage1',
        src: 'source1',
      },
    },
    allIds: ['backImage1'],
  },
  settings: {
    cardsQuantity: 4,
  },
};

describe('GameBoard generator', () => {
  describe('generateCards', () => {
    it('should return the array of cards', () => {
      const cards = generateCards(state, false);
      expect(cards).toMatchSnapshot();
    });
  });
});
