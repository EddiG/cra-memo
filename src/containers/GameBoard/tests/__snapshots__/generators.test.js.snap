// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBoard generator generateCards should return the array of cards 1`] = `
Array [
  Object {
    "backImageId": "backImage1",
    "frontImageId": "frontImage1",
    "id": "card1",
    "isGuessed": false,
    "isOpened": false,
    "pairCardId": "card3",
  },
  Object {
    "backImageId": "backImage1",
    "frontImageId": "frontImage1",
    "id": "card3",
    "isGuessed": false,
    "isOpened": false,
    "pairCardId": "card1",
  },
  Object {
    "backImageId": "backImage1",
    "frontImageId": "frontImage2",
    "id": "card2",
    "isGuessed": false,
    "isOpened": false,
    "pairCardId": "card4",
  },
  Object {
    "backImageId": "backImage1",
    "frontImageId": "frontImage2",
    "id": "card4",
    "isGuessed": false,
    "isOpened": false,
    "pairCardId": "card2",
  },
]
`;
